### DISCUSSION TITLE
Why `os.remove` works with `.py` but fails to work with `.exe` in case of deleting the file being executed

### DISCUSSION THREAD

**Discussion by sourabhkv** [pyinstaller/pyinstaller#7639] [PyInstaller] [answered]
I have a python file named as `test.py`
```python
import os
print("running before deletion")
try:
    os.remove(__file__)     #for .py file
except:
    os.remove("test.exe")   #for .exe file
print("running after deletion")
print(__file__)
```
Running for .py
![image](https://github.com/pyinstaller/pyinstaller/assets/55890376/731d9686-95e4-41cb-9e23-61f0065197a8)

Running for .exe
![image](https://github.com/pyinstaller/pyinstaller/assets/55890376/126c6107-cf6b-4751-a2e0-210bdd00d19a)
The except block also fails because of permission error but why this didn't appear in case of .py?

I am working on an application which involves updating the application , basically my application(.exe) searches in GitHub for latest version if it finds newer version it asks for permission(y/N) from user but I am unable to delete executable while it is executing but somehow this works with py files.

Can anyone answer why it's like that ? and also help how to update my application (.exe) ?

**Comment by bwoodsend** [comment, marked as answer]
Because when Windows is running an executable, it keeps its file handle open whereas Python reads a `.py` file into memory and closes the file before executing it. For an executable to modify itself, it needs to create a subprocess that does all the work, disown the subprocess so that the subprocess won't exit when the application quits, then quit so that the subprocess is allowed to modify the executable.

I always wished that self updating applications didn't exist but if you don't feel that way I suggest you use [pyupdater](https://pypi.org/project/PyUpdater/).

**Reply by sourabhkv** [reply]
Thanks
